projects/FW_104/src/fw_104_can.c

```diff
@@ -0,0 +1,102 @@
+#include "can.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+
+#include "fw_104_can.h"
+
+#define CAN_SEND_TIME_MS 1000
+
+StatusCode prv_can_callback(const CanMessage *msg, void *context, CanAckStatus *ack_reply) {
+  // Taken from smoke_can
+  LOG_DEBUG("Received a message!\n");
+  printf("Data:\n\t");
+  for (uint8_t i = 0; i < msg->dlc; i++) {
+    uint8_t byte = (msg->data >> (i * 8)) & 0xFF;
+    printf("0x%x ", byte);
+  }
+  printf("\n");
+  return STATUS_CODE_OK;
+}
+
+void prv_can_transmit_A(SoftTimerId timer_id, void *context) {
+  uint8_t messageData = 5;
+  CanMessage message = {
+    .source_id = 0x1, .msg_id = 0x1, .data_u8 = { messageData }, .type = CAN_MSG_TYPE_DATA, .dlc = 8
+  };
+  can_transmit(&message, NULL);
+
+  soft_timer_start_millis(CAN_SEND_TIME_MS, prv_can_transmit_A, NULL, NULL);
+}
+
+void prv_can_transmit_B(SoftTimerId timer_id, void *context) {
+  uint8_t messageData = 5;
+  CanMessage message = {
+    .source_id = 0x1, .msg_id = 0x1, .data_u8 = { messageData }, .type = CAN_MSG_TYPE_DATA, .dlc = 8
+  };
+  can_transmit(&message, NULL);
+
+  soft_timer_start_millis(CAN_SEND_TIME_MS, prv_can_transmit_B, NULL, NULL);
+}
+
+void prv_write_A_message() {
+  CanStorage storage;
+
+  CanSettings settings = {
+    .device_id = 0xA,
+    .bitrate = CAN_HW_BITRATE_500KBPS,
+    // Not really sure what to put here, following smoke_can
+    .rx_event = 0,
+    .tx_event = 1,
+    .fault_event = 2,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+    .loopback = true,
+  };
```

Comment:
A few points:
* you only really need to initialize CAN once, it's usually done in main but for homework purposes, whatever
* `loopback = true` tells the CAN library to receive its own messages, I think the idea with the "run it on two terminals" thing was for loopback to be false so you see the messages and the ACKs travelling across the virtual CAN bus between the two instances of your program
* the tx/rx/fault_event are event IDs. I'll talk about the event queue on Saturday but those events are used for driving the CAN library (tx is raised when a message is being tx'd (transmitted), rx is raised when a message was rx'd (received), and fault is raised on a fault). Usually you should define them in an enum kinda like this:
```c
typedef enum {
  FW_104_CAN_RX_EVENT = 0,
  FW_104_CAN_TX_EVENT,
  FW_104_CAN_FAULT_EVENT,
  NUM_FW_104_CAN_EVENTS,
} Fw104CanEvent;
```
You can see examples of these event enums in e.g. solar_events.h.